File: client/lib/services/util/extensions/order_type_extensions.dart
import 'package:solidtrade/data/models/enums/client_enums/order_type.dart';
import 'package:solidtrade/data/models/enums/entity_enums/enter_or_exit_position_type.dart';
import 'package:solidtrade/data/models/enums/shared_enums/buy_or_sell.dart';

extension OrderTypeExtension on OrderType {
  String get name {
    switch (this) {
      case OrderType.market:
        return "Market";
      case OrderType.stop:
        return "Stop";
      case OrderType.limit:
        return "Limit";
    }
  }

  String get fullName {
    switch (this) {
      case OrderType.market:
        return "Market order";
      case OrderType.stop:
        return "Stop order";
      case OrderType.limit:
        return "Limit order";
    }
  }

  EnterOrExitPositionType toEnterOrExitPosition(BuyOrSell buyOrSell) {
    switch (this) {
      case OrderType.stop:
        switch (buyOrSell) {
          case BuyOrSell.buy:
            return EnterOrExitPositionType.buyStopOrder;
          case BuyOrSell.sell:
            return EnterOrExitPositionType.sellStopOrder;
        }
      case OrderType.limit:
        switch (buyOrSell) {
          case BuyOrSell.buy:
            return EnterOrExitPositionType.buyLimitOrder;
          case BuyOrSell.sell:
            return EnterOrExitPositionType.sellLimitOrder;
        }
      case OrderType.market:
        throw "Did not expect market order type";
    }
  }
}

File: server/Application/Extensions/OrderTypeExtensions.cs
using Application.Models.Dtos.TradeRepublic;
using Domain.Enums;

namespace Application.Extensions;

public static class OrderTypeExtensions
{
    public static bool IsOrderFulfilled(this OrderType type, TradeRepublicProductPriceResponseDto value, decimal orderPrice)
    {
        return type switch
        {
            // Current has to be below
            OrderType.BuyLimitOrder => orderPrice >= value.Ask.Price,
            // Current has to be above
            OrderType.BuyStopOrder => value.Ask.Price >= orderPrice,
            // Current has to be above (take profit)
            OrderType.SellLimitOrder => orderPrice <= value.Bid.Price,
            // Current has to be below (stop loss)
            OrderType.SellStopOrder => value.Bid.Price <= orderPrice,
        };
    }
    
    public static BuyOrSell IsBuyOrSell(this OrderType type)
    {
        return type switch
        {
            OrderType.BuyLimitOrder => BuyOrSell.Buy,
            OrderType.BuyStopOrder => BuyOrSell.Buy,
            OrderType.SellLimitOrder => BuyOrSell.Sell,
            OrderType.SellStopOrder => BuyOrSell.Sell,
        };
    }
    
    public static string UserFriendlyFullName(this OrderType type)
    {
        return type switch
        {
            OrderType.BuyLimitOrder => "buy limit",
            OrderType.BuyStopOrder => "buy stop",
            OrderType.SellLimitOrder => "take profit",
            OrderType.SellStopOrder => "stop loss",
        };
    }
}